sklearn/datasets/rcv1.py

```diff
@@ -0,0 +1,201 @@
+        The array has 0.16% of non zero values.
+
+    dataset.target : scipy csr array, shape (804414, 103)
+        Each sample has a value of 1 in its categories, and 0 in others.
+        The array has 3.15% of non zero values.
+
+    dataset.sample_id : numpy array, shape (804414,)
+        Identification number of each sample, as ordered in dataset.data.
+
+    dataset.categories : list of string, length (103)
+        Names of each category, as ordered in dataset.target.
+
+    dataset.DESCR : string
+        Description of the RCV1 dataset.
+
+    """
+    N_SAMPLES = 804414
+    N_FEATURES = 47236
+    N_CATEGORIES = 103
+
+    data_home = get_data_home(data_home=data_home)
+    if sys.version_info[0] == 3:
+        # The zlib compression format use by joblib is not compatible when
+        # switching from Python 2 to Python 3, let us use a separate folder
+        # under Python 3:
+        dir_suffix = "-py3"
+    else:
+        # Backward compat for Python 2 users
+        dir_suffix = ""
+    rcv1_dir = join(data_home, "RCV1" + dir_suffix)
+    samples_path = join(rcv1_dir, "samples")
+    targets_path = join(rcv1_dir, "targets")
+    sample_id_path = join(rcv1_dir, "sample_id")
+    categories_path = join(rcv1_dir, "categories")
+    available = exists(samples_path)
+
+    if download_if_missing and not available:
+        _mkdirp(rcv1_dir)
+
+        # load data (X) and sample_id
+        file_urls = ["%s_test_pt%d.dat.gz" % (URL, i) for i in range(4)]
+        file_urls.append("%s_train.dat.gz" % URL)
+        files = []
+        for file_url in file_urls:
+            logger.warning("Downloading %s" % file_url)
+            f = BytesIO(urlopen(file_url).read())
+            files.append(GzipFile(fileobj=f))
+
+        Xy = load_svmlight_files(files, n_features=N_FEATURES)
+
+        # Training data is before testing data
+        X = sp.vstack([Xy[8], Xy[0], Xy[2], Xy[4], Xy[6]]).tocsr()
+        sample_id = np.hstack((Xy[9], Xy[1], Xy[3], Xy[5], Xy[7]))
+        sample_id = sample_id.astype(np.int32)
+
+        # load target (y) and sample_id_bis
+        logger.warning("Downloading %s" % URL_topics)
+        f = BytesIO(urlopen(URL_topics).read())
+        f = GzipFile(fileobj=f, mode='r')
+
+        n_cat = -1
+        n_doc = -1
+        doc_previous = -1
+        y = np.zeros((N_SAMPLES, N_CATEGORIES), dtype=np.int8)
+        sample_id_bis = np.zeros(N_SAMPLES, dtype=np.int32)
+        category_names = {}
+        for line in f.readlines():
+            if len(line.split(" ")) > 2:
+                cat, doc, _ = line.split(" ")
+                if cat not in category_names:
+                    n_cat += 1
+                    category_names[cat] = n_cat
+
+                doc = int(doc)
+                if doc != doc_previous:
+                    doc_previous = doc
+                    n_doc += 1
+                    sample_id_bis[n_doc] = doc
+                y[n_doc, category_names[cat]] = 1
```

Comment:
This could be probably replace by some numpy line.

Something like

```python

category, sample_id = np.loadtxt(f, unpack=True, usecols=(0, 1), dtype=object)
sample_id = sample_id.astype(int)
_, category = np.unique(category, return_inverse=True)
y = coo_matrix((np.one_like(category), (sample_id, category)),
               shape=(N_SAMPLES, N_CATEGORIES)).tocsr()

```

Note that I haven't tested the previous code and also that loadtxt can decompress `.gz` format